addons/udes_stock/tests/test_push_from_drop.py

```diff
@@ -0,0 +1,253 @@
+            "location_id": cls.location_qc_zone.id,
+            "picking_type_id": cls.picking_type_putaway.id,
+            "u_push_on_drop": True,
+            "action": "push",
+        })
+
+    def setUp(self):
+        """
+        Ensure all tests start with push_from_drop set to True
+        """
+        super(PushFromDropBase, self).setUp()
+        self.push_putaway.u_push_on_drop = True
+        self.push_damaged_putaway.u_push_on_drop = True
+
+
+class TestGetPathFromLocation(PushFromDropBase):
+    """Tests for the stock.rule.get_path_from_location method in
+    isolation"""
+    def test01_no_route(self):
+        """Test that get_path_from_location returns an empty recordset when no
+        matching path is found"""
+        Push = self.env['stock.rule']
+        self.push_damaged_putaway.unlink()
+        res = Push.get_path_from_location(self.received_damaged_location)
+        self.assertFalse(res)
+
+    def test02_one_route(self):
+        """Test that get_path_from_location returns a single matching path when
+        one exists"""
+        Push = self.env['stock.rule']
+        res = Push.get_path_from_location(self.received_damaged_location)
+        self.assertEqual(res, self.push_damaged_putaway)
+
+    def test03_many_routes(self):
+        """Test that get_path_from_location returns a single matching path when
+        more than one possible match exists, and the returned value is the one
+        for the closest parent of the search location.
+        """
+        Push = self.env['stock.rule']
+        input_zone = self.env.ref('stock.stock_location_company')
+
+        # create new push from the Input zone.
+        Push.create({
+            "name": "TestPushMultiRoute",
+            "route_id": self.route_in.id,
+            "sequence": 40,
+            "location_src_id": input_zone.id,
+            "location_id": self.location_qc_zone.id,
+            "picking_type_id": self.picking_type_putaway.id,
+            "u_push_on_drop": True,
+            "action": "push",
+        })
+        # Damaged route found correctly
+        res = Push.get_path_from_location(self.received_damaged_location)
+        self.assertEqual(res, self.push_damaged_putaway)
+
+        # Non-damaged route found correctly
+        res = Push.get_path_from_location(self.received_location)
+        self.assertEqual(res, self.push_putaway)
+
+
+class TestCreateMovesForPush(PushFromDropBase):
+    """Tests for the stock.move._create_moves_for_push method in
+    isolation"""
+    def setUp(self):
+        """Create a picking to work with for tests"""
+        super(TestCreateMovesForPush, self).setUp()
+        products_info = [{'product': self.apple, 'uom_qty': 10},
+                         {'product': self.banana, 'uom_qty': 99}]
+        self.goods_in = self.create_picking(self.picking_type_goods_in,
+                                            products_info=products_info)
+        self.goods_in.action_confirm()
```

Comment:
Don't need these -can pass them as kwargs into create_picking

Reply:
I just want to check I understand; I should pass `products_info` to create picking and not `products_info = products_info`?

Reply:
ah, so action_confirm() and action_assign() are triggered inside create_picking function, so you can do 
```self.create_picking(self.picking_type_goods_in,
                                            products_info=products_info, confirm=True, assign=True)
```
instead of calling them separately. Still pass products_info as you have though